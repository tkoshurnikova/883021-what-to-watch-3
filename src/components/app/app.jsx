import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Switch, Route, BrowserRouter} from "react-router-dom";
import Main from "../main/main.jsx";
import MoviePage from "../movie-page/movie-page.jsx";

class App extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      clickedCard: null
    };

    this._onCardClick = this._onCardClick.bind(this);
  }

  _onCardClick(film) {
    this.setState({
      clickedCard: film
    });
  }

  _renderApp() {
    const {filmName, filmGenre, filmReleaseDate, films} = this.props;
    const {clickedCard} = this.state;

    if (clickedCard) {
      return (
        <MoviePage
          film={clickedCard}
        />
      );
    }

    return (
      <Main
        filmName={filmName}
        filmGenre={filmGenre}
        filmReleaseDate={filmReleaseDate}
        films={films}
        onCardClick={this._onCardClick}
      />
    );
  }

  render() {
    const {clickedCard} = this.state;
    const {films} = this.props;
    const film = (clickedCard) ? clickedCard : films[0];

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/movie-page">
            <MoviePage
              film={film}
            />
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

App.propTypes = {
  filmName: PropTypes.string.isRequired,
  filmGenre: PropTypes.string.isRequired,
  filmReleaseDate: PropTypes.number.isRequired,
  films: PropTypes.array.isRequired
};

export default App;
