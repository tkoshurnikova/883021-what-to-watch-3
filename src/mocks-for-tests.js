export const FilmDetails = {
  title: `Whispering Mist`,
  genre: `Comedy`,
  year: 1995,
  video: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
};

export const films = [
  {
    "title": `The No Flowers`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `Rings of Cloud`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `The Sword's Memory`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `The Voyagers of the Words`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `Shores in the Heart`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `Missing Illusion`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `The Unwilling Man`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  },
  {
    "title": `Teacher in the Hunter`,
    "image": `http://placehold.it/280x175`,
    "description": `qwerty`,
    "rating": 1.0,
    "numberOfVotes": 1982,
    "director": `Taika Di Caprio`,
    "actors": [`Ashley Cooper`, `Ramona Flowers`],
    "genre": `Dramas`,
    "year": 1900,
    "preview": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "video": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    "background_color": `#ffffff`,
    "background_image": `some image`,
    "poster_image": `some image`
  }
];
