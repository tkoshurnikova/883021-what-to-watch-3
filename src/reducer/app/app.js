import {Genre, CARDS_TO_SHOW} from "../../const";
import {extend} from "../../utils";

const initialState = {
  genre: Genre.ALL_GENRES,
  cardsToShow: CARDS_TO_SHOW,
  error: false
};

export const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  SHOW_MORE_FILMS: `SHOW_MORE_FILMS`,
  RESET_CARDS_COUNT: `RESET_CARDS_COUNT`,
  REQUEST_FAIL: `REQUEST_FAIL`
};

export const ActionCreator = {
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    payload: genre
  }),
  showMoreFilms: () => ({
    type: ActionType.SHOW_MORE_FILMS,
    payload: CARDS_TO_SHOW
  }),
  resetCardsCount: () => ({
    type: ActionType.RESET_CARDS_COUNT,
  }),
  requestFail: () => ({
    type: ActionType.REQUEST_FAIL,
    payload: true
  })
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      return extend(state, {
        genre: action.payload
      });

    case ActionType.SHOW_MORE_FILMS:
      return extend(state, {
        cardsToShow: state.cardsToShow + action.payload
      });

    case ActionType.RESET_CARDS_COUNT:
      return extend(state, {
        cardsToShow: CARDS_TO_SHOW
      });

    case ActionType.REQUEST_FAIL:
      return extend(state, {
        error: action.payload
      });
  }

  return state;
};
