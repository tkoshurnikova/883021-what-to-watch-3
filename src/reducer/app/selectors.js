import NameSpace from "../name-space.js";

const NAME_SPACE = NameSpace.APP;

export const getGenre = (state) => {
  return state[NAME_SPACE].genre;
};

export const getCardsToShow = (state) => {
  return state[NAME_SPACE].cardsToShow;
};

export const getErrorStatus = (state) => {
  return state[NAME_SPACE].error;
};
