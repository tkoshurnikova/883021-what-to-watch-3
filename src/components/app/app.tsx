import * as React from "react";
import {Switch, Route, HashRouter} from "react-router-dom";
import {connect} from "react-redux";
import history from "../../history";

import Main from "../main/main";
import SignIn from "../sign-in/sign-in";
import MoviePage from "../movie-page/movie-page";
import FullscreenPlayer from "../fullscreen-player/fullscreen-player";
import AddReview from "../add-review/add-review";
import MyList from "../my-list/my-list";
import Error from "../error/error";
import PrivateRoute from "../private-route/private-route";

import withActiveItem from "../../hocs/with-active-item/with-active-item";
import withFullscreenVideo from "../../hocs/with-fullscreen-video/with-fullscreen-video";
import withReviewValidation from "../../hocs/with-review-validation/with-review-validation";

import {getFilms, getPromoFilm} from "../../reducer/data/selectors";
import {getErrorStatus} from "../../reducer/app/selectors";
import {Operation as UserOperation} from "../../reducer/user/user";
import {Operation as DataOperations} from "../../reducer/data/data";

import {AppRoute} from "../../const";
import {getFilmByID} from "../../utils";
import {Film} from "../../types";

const WrappedMoviePage = withActiveItem(MoviePage);
const WrappedFullscreenPlayer = withFullscreenVideo(FullscreenPlayer);
const WrappedAddReview = withReviewValidation(AddReview);

interface Props {
  films: Film[];
  promoFilm: Film;
  login: () => void;
  sendReview: () => void;
  error: boolean;
}

const App: React.FunctionComponent<Props> = (props: Props) => {

  const {
    films,
    promoFilm,
    sendReview,
    login,
    error
  } = props;

  if (error) {
    return <Error/>;
  }

  return (
    <HashRouter>
      <Switch>
        <Route
          exact
          path={AppRoute.MAIN}
          render={() => {
            return (
              <Main
                PromoFilm={promoFilm}
              />
            );
          }}
        />
        <Route
          exact
          path={AppRoute.LOGIN}
          render={() => {
            return (
              <SignIn onSubmit={login} />
            );
          }}
        />
        <Route
          exact
          path={`${AppRoute.FILMS}/:id`}
          render={(routeProps) => {
            const {id} = routeProps.match.params;
            const film = getFilmByID(films, id);
            return (
              <WrappedMoviePage
                film={film}
                chosenFilm={film}
              />
            );
          }}
        />
        <PrivateRoute
          exact
          path={`${AppRoute.FILMS}/:id/review`}
          render={(routeProps) => {
            const {id} = routeProps.match.params;
            const film = getFilmByID(films, id);
            return (
              <WrappedAddReview
                onSubmit={sendReview}
                film={film}
              />
            );
          }}
        />
        <Route
          exact
          path={`${AppRoute.PLAYER}/:id`}
          render={(routeProps) => {
            const {id} = routeProps.match.params;
            const film = getFilmByID(films, id);
            return (
              <WrappedFullscreenPlayer
                film={film}
              />
            );
          }}
        />
        <PrivateRoute
          exact
          path={AppRoute.MY_LIST}
          render={() => {
            return (
              <MyList />
            );
          }}
        />
      </Switch>
    </HashRouter>
  );
};

const mapStateToProps = (state) => ({
  films: getFilms(state),
  promoFilm: getPromoFilm(state),
  error: getErrorStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  login(authData) {
    dispatch(UserOperation.login(authData));
  },
  sendReview(authData, id) {
    dispatch(DataOperations.sendReview(authData, id));
  }
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
