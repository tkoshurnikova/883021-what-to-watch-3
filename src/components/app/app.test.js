import React from "react";
import renderer from "react-test-renderer";
import {App} from "./app.jsx";
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import {films, FilmDetails} from "../../mocks-for-tests";

const mockStore = configureMockStore([]);
let store = mockStore({
  APP: {
    genre: ``,
    cardsToShow: 8
  },
  DATA: {
    films,
    promoFilm: FilmDetails
  }
});

it(`Render App`, () => {
  const tree = renderer
    .create(
        <Provider store={store}>
          <App
            promoFilm={FilmDetails}
            films={films}
            onCardClick={() => {}}
            clickedCard={films[0]}
            onPlayOrExitButtonClick={() => {}}
          />
        </Provider>, {
          createNodeMock: () => {
            return {};
          }
        })
    .toJSON();

  expect(tree).toMatchSnapshot();
});
