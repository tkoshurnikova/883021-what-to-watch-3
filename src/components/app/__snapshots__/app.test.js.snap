// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/bg-the-grand-budapest-hotel.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <img
            alt="User avatar"
            height="63"
            src="img/avatar.jpg"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Whispering Mist
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              1995
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Comedies
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Crime
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Documentary
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Dramas
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Horror
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Kids & Family
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Romance
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Sci-Fi
          </a>
        </li>
        <li
          className="catalog__genres-item"
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Thrillers
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="The No Flowers"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              The No Flowers
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Rings of Cloud"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              Rings of Cloud
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="The Sword's Memory"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              The Sword's Memory
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="The Voyagers of the Words"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              The Voyagers of the Words
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Shores in the Heart"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              Shores in the Heart
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Missing Illusion"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              Missing Illusion
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="The Unwilling Man"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              The Unwilling Man
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <img
              alt="Teacher in the Hunter"
              height="175"
              src="http://placehold.it/280x175"
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page"
              onClick={[Function]}
            >
              Teacher in the Hunter
            </a>
          </h3>
        </article>
      </div>
      <div
        className="catalog__more"
      >
        <button
          className="catalog__button"
          type="button"
        >
          Show more
        </button>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
